### PR TITLE
Implement ChargedData

### PR DESCRIPTION
**Pull request by clienthax**
Implementation of `REPLACE_WITH_DATA_MANIPULATOR`.
- [x] Registration of field getters and setters
- [x] Accurately used the appropriate `AbstractData` implementation

Implementation of `REPLACE_WITH_IMMUTABLE_MANIPULATOR`
- [x] Accurately extended the appropriate `AbstractImmutableData` implementation
- [x] Accurately instantiating final instance fields with an `ImmutableValue` counter part for any value getters
- [x] If necessary, creating a new `ImmutableValue` for a value that should not be cached, or, using the existing caching utils.

Implementation of the `VALUE_PROCESSOR`(s):
- [x] Appropriately extended `AbstractSpongeValueProcessor`
- [x] Individual processors for each source type possible (one for `ItemStack`/`TileEntity`/`Entity` as necessary).

Registration:
- [x] Registered the `Key` correctly by `FIELD_NAME` in the `KeyRegistry`
- [x] Registered the `DataProcessor`s and `DataManipulatorBuilder` in `SpongeSerializationRegistry`
- [x] Registered the `ValueProcessor`s in the `SpongeSerializationRegistry`

``` java
import com.google.inject.Inject;
import org.spongepowered.api.Game;
import org.spongepowered.api.data.DataContainer;
import org.spongepowered.api.data.key.Keys;
import org.spongepowered.api.data.manipulator.DataManipulatorRegistry;
import org.spongepowered.api.data.manipulator.mutable.entity.AgentData;
import org.spongepowered.api.data.manipulator.mutable.entity.ChargedData;
import org.spongepowered.api.entity.Entity;
import org.spongepowered.api.entity.EntityTypes;
import org.spongepowered.api.entity.living.player.Player;
import org.spongepowered.api.event.Listener;
import org.spongepowered.api.event.cause.Cause;
import org.spongepowered.api.event.game.state.GameInitializationEvent;
import org.spongepowered.api.plugin.Plugin;
import org.spongepowered.api.service.persistence.InvalidDataException;
import org.spongepowered.api.text.Texts;
import org.spongepowered.api.util.command.CommandException;
import org.spongepowered.api.util.command.CommandResult;
import org.spongepowered.api.util.command.CommandSource;
import org.spongepowered.api.util.command.args.CommandContext;
import org.spongepowered.api.util.command.args.GenericArguments;
import org.spongepowered.api.util.command.spec.CommandExecutor;
import org.spongepowered.api.util.command.spec.CommandSpec;

import java.util.Optional;

@Plugin(name = "ChargedTest", id="test2")
public class ChargedTest {

    @Listener
    public void onServerStart(GameInitializationEvent event)
    {
        CommandSpec speedCommandSpec = CommandSpec.builder()
                .description(Texts.of("Test Command"))
                .permission("test.use")
                .arguments(GenericArguments.optional(GenericArguments.onlyOne(GenericArguments.bool(Texts.of("charged")))))
                .executor(new ChargedValueProcessorCommand())
                .build();
        event.getGame().getCommandDispatcher().register(this, speedCommandSpec, "chargedcreeper");
    }

    @Inject
    public Game game;

    public class ChargedValueProcessorCommand implements CommandExecutor {
        @Override
        public CommandResult execute(CommandSource src, CommandContext args) throws CommandException {
            if(!(src instanceof Player))
                return CommandResult.empty();
            Player player = (Player) src;
            boolean aiEnabled = (Boolean) args.getOne("charged").get();

            Optional<Entity> expOrb = player.getWorld().createEntity(EntityTypes.CREEPER, player.getLocation().getPosition());
            Entity entity = expOrb.get();
            player.getWorld().spawnEntity(entity, Cause.empty());
            entity.offer(Keys.CREEPER_CHARGED, aiEnabled);


            final ChargedData origData = entity.get(ChargedData.class).get();
            final DataContainer container = origData.toContainer();
            final DataManipulatorRegistry registry = game.getRegistry().getManipulatorRegistry();
            ChargedData chargedData = registry.getBuilder(ChargedData.class).get().build(container).get();
            if(chargedData.compareTo(origData) != 0) {
                throw new InvalidDataException();
            }
            if(!origData.equals(chargedData)) {
                throw new InvalidDataException();
            }


            return CommandResult.success();
        }
    }

}
```

### CODE REVIEW

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeChargedData.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	}
+
+	@Override
+	protected ImmutableValue<?> getValueGetter() {
+		return charged();
+	}
+
+	@Override
+	public ImmutableValue<Boolean> charged() {
+		return ImmutableSpongeValue.cachedOf(Keys.CREEPER_CHARGED, false, this.value);
```
Comment (ZephireNZ) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeChargedData.java]: Make this a final field - it never changes once the data is created, so there's no point trying to find the data each time.

File: src/main/java/org/spongepowered/common/data/processor/value/entity/IsCreeperChargedValueProcessor.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+import org.spongepowered.api.data.value.ValueContainer;
+import org.spongepowered.api.data.value.immutable.ImmutableValue;
+import org.spongepowered.api.data.value.mutable.Value;
+import org.spongepowered.common.data.processor.common.AbstractSpongeValueProcessor;
+import org.spongepowered.common.data.value.immutable.ImmutableSpongeValue;
+import org.spongepowered.common.data.value.mutable.SpongeValue;
+
+import java.util.Optional;
+
+public class IsCreeperChargedValueProcessor extends AbstractSpongeValueProcessor<EntityCreeper, Boolean, Value<Boolean>> {
```
Comment (ZephireNZ) [inline src/main/java/org/spongepowered/common/data/processor/value/entity/IsCreeperChargedValueProcessor.java]: Why not `CreeperChargedValueProcessor` to match the key?

Comment (ZephireNZ) [inline src/main/java/org/spongepowered/common/data/processor/value/entity/IsCreeperChargedValueProcessor.java]: Or even just `ChargedValueProcessor`, as creeper (afaik) are the only thing that can be charged.

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeChargedData.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import org.spongepowered.common.data.manipulator.immutable.common.AbstractImmutableBooleanData;
+import org.spongepowered.common.data.manipulator.mutable.entity.SpongeChargedData;
+import org.spongepowered.common.data.value.immutable.ImmutableSpongeValue;
+
+public class ImmutableSpongeChargedData extends AbstractImmutableBooleanData<ImmutableChargedData, ChargedData> implements ImmutableChargedData {
+
+	private final ImmutableValue<Boolean> cachedValue;
+
+	public ImmutableSpongeChargedData(Boolean value) {
+		super(ImmutableChargedData.class, value, Keys.CREEPER_CHARGED, SpongeChargedData.class);
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeChargedData.java]: Same here, replace the tabs with spaces

**Pull request closed without merging**